packages/hardhat/test/MyUSDEngine.ts

```diff
@@ -0,0 +1,606 @@
+//
+// This script executes when you run 'yarn test'
+//
+
+import { ethers } from "hardhat";
+import { expect } from "chai";
+import { MyUSD, DEX, MyUSDEngine, Oracle, MyUSDStaking, RateController } from "../typechain-types";
+import { SignerWithAddress } from "@nomicfoundation/hardhat-ethers/signers";
+
+describe("🚩 Stablecoin Challenge 🤓", function () {
+  let myUSDToken: MyUSD;
+  let dex: DEX;
+  let myUSDEngine: MyUSDEngine;
+  let oracle: Oracle;
+  let staking: MyUSDStaking;
+  let rateController: RateController;
+  let owner: SignerWithAddress;
+  let user1: SignerWithAddress;
+  let user2: SignerWithAddress;
+
+  const collateralAmount = ethers.parseEther("10");
+  const borrowAmount = ethers.parseEther("5000");
+
+  beforeEach(async function () {
+    await ethers.provider.send("hardhat_reset", []);
+    [owner, user1, user2] = await ethers.getSigners();
+
+    // Get the deployer's current nonce
+    const deployerNonce = await ethers.provider.getTransactionCount(owner.address);
+
+    // Calculate future addresses based on nonce
+    const futureStakingAddress = ethers.getCreateAddress({
+      from: owner.address,
+      nonce: deployerNonce + 4, // +4 because it will be our fifth deployment
+    });
+
+    const futureEngineAddress = ethers.getCreateAddress({
+      from: owner.address,
+      nonce: deployerNonce + 5, // +5 because it will be our sixth deployment
+    });
+
+    // Deploy RateController first
+    const RateControllerFactory = await ethers.getContractFactory("RateController");
+    rateController = await RateControllerFactory.deploy(futureEngineAddress, futureStakingAddress);
+
+    // Deploy MyUSD with future addresses
+    const MyUSDFactory = await ethers.getContractFactory("MyUSD");
+    myUSDToken = await MyUSDFactory.deploy(futureEngineAddress, futureStakingAddress);
+
+    // Deploy DEX
+    const DEXFactory = await ethers.getContractFactory("DEX");
+    dex = await DEXFactory.deploy(await myUSDToken.getAddress());
+
+    // Deploy Oracle
+    const OracleFactory = await ethers.getContractFactory("Oracle");
+    oracle = await OracleFactory.deploy(await dex.getAddress());
+
+    // Deploy MyUSDStaking
+    const MyUSDStakingFactory = await ethers.getContractFactory("MyUSDStaking");
+    staking = await MyUSDStakingFactory.deploy(
+      await myUSDToken.getAddress(),
+      futureEngineAddress,
+      await rateController.getAddress(),
+    );
+
+    // Finally deploy the MyUSDEngine at the predicted address
+    const MyUSDEngineFactory = await ethers.getContractFactory("MyUSDEngine");
```

Comment:
We will need to do [this special dance](https://github.com/scaffold-eth/se-2-challenges/blob/8fe0977aaaf73a95183e9c9e012c809277df8ce3/extension/packages/hardhat/test/OverCollateralizedLending.ts#L38) with making this to reference a different contract in order to allow the SRE autograder to work. 

Once you have added this logic it is simple to test with this command:
`CONTRACT_ADDRESS=0xtest yarn test`

And just make sure you have a duplicate of your contract with the name `download-0xtest.sol` with one method slightly messed up so that you can prove it is picking up the correct contract (not the default).

Reply:
It will not be necessary to do this for the Staking contract. Students will only be tested on their Egnine contract (but it is still good that we have tests in place for the staking contract so they can complete it locally).